### PR TITLE
adding support for eligible_users flag in llm

### PR DESCRIPTION
**Pull request by joker2411**
## Description of the change

> Adding support for eligible_users conditions in case of LLM. Ticket [link](https://linear.app/rudderstack/issue/PRML-617/add-eligible-users-flag-to-llm-pynative-models).

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: samples/llm_model/models/profiles.yaml [hunk 1]
```diff
@@ -17,17 +17,20 @@ models:
...
         - user.Var('name')
       llm_model_name: "mistral-7b"
+      run_for_top_k_distinct: 10
   
   - name: llm_profession_prediction
     model_type: llm_model
     model_spec:
       prompt: "Following are the titles and links to top 100 pages on RudderStack website in JSON format : {sql_inputs[0]} . You are a sales development representative. The input field is a job title entered by a person. Map it to a single value from the following list of standardised job titles: [Software Engineer, Data Analyst, ML Engineer, Solutions Architect, Others]. Return only standardised title, nothing else. {var_inputs[0]}"
       entity_key: user
+      eligible_users: "{var_inputs[0]} is not null"
```
Comment (dpatchigolla) [inline samples/llm_model/models/profiles.yaml:27]: Can't we use the user.Var() syntax here directly?

File: src/predictions/profiles_mlcorelib/py_native/llm.py [hunk 2]
```diff
@@ -23,15 +24,18 @@ class LLMModel(BaseModelType):
         "required": ["prompt", "var_inputs"] + EntityKeyBuildSpecSchema["required"],
         "additionalProperties": False,
     }
-    DEFAULT_SQL_INPUTS = []
-    DEFAULT_LLM_MODEL = "llama2-70b-chat"
+    DEFAULT_VALUES = {
+        "eligible_users": "1=1",
```
Comment (dpatchigolla) [inline src/predictions/profiles_mlcorelib/py_native/llm.py:28]: Can't we skip it if we don't have eligible users param? may be use an empty string?

File: src/predictions/profiles_mlcorelib/py_native/llm.py [hunk 3]
```diff
@@ -23,15 +24,18 @@ class LLMModel(BaseModelType):
...
+    }
 
     def __init__(self, build_spec: dict, schema_version: int, pb_version: str) -> None:
         super().__init__(build_spec, schema_version, pb_version)
-        if "sql_inputs" not in self.build_spec:
-            self.build_spec["sql_inputs"] = self.DEFAULT_SQL_INPUTS
-        if self.build_spec["llm_model_name"] == None:
-            self.build_spec["llm_model_name"] = self.DEFAULT_LLM_MODEL
+        for key in ("eligible_users", "sql_inputs", "llm_model_name"):
+            if key not in self.build_spec or self.build_spec[key] == None:
```
Comment (dpatchigolla) [inline src/predictions/profiles_mlcorelib/py_native/llm.py:37]: minor suggestion:
`== None` can be replaced by `is None`, more pythonic.

File: src/predictions/profiles_mlcorelib/py_native/llm.py [hunk 4]
```diff
@@ -99,6 +109,53 @@ def describe(self, this: WhtMaterial):
         return self.sql, ".sql"
 
     def register_dependencies(self, this: WhtMaterial):
+        def _get_index_list(text_input):
+            text_input_replaced = text_input.replace(
+                "'", "''", -1
+            )  # This is to escape single quotes
```
Comment (dpatchigolla) [inline src/predictions/profiles_mlcorelib/py_native/llm.py:115]: We should always expect the prompts to be in double quotes, not single quotes?

File: src/predictions/profiles_mlcorelib/py_native/llm.py [hunk 5]
```diff
@@ -99,6 +109,53 @@ def describe(self, this: WhtMaterial):
...
+            return (
+                re.findall(r"{(\w+)\[(\d+)\]}", text_input_replaced),
+                text_input_replaced,
+            )
+
+        def _replace_placeholders(
```
Comment (dpatchigolla) [inline src/predictions/profiles_mlcorelib/py_native/llm.py:121]: Can you add a few test cases for this, trying out a few scenarios:
1. Where var inputs is an empty list
2. Where the prompt does not have any input references
3. Where prompt has only var_input ref but no sql input ref
4. Eligible users is empty
5. Eligible users is not empty
Plz add anything else you seem may be risky

File: src/predictions/profiles_mlcorelib/py_native/llm.py [hunk 6]
```diff
@@ -152,18 +213,22 @@ def register_dependencies(self, this: WhtMaterial):
                 {{% macro selector_sql() %}}
                     {{% set entityVarTable = {var_table_ref} %}}
                     -- Common Table Expression (CTE) to get the distinct values of specified columns based on their frequency
-                    WITH top_k_distinct_attribute AS (
-                        SELECT {joined_columns}, COUNT(*) AS frequency
+                    WITH filtered_entity_var_table AS(
+                        SELECT *
                         FROM {{{{entityVarTable}}}}
+                        WHERE {eligible_users_prompt_replaced}
```
Comment (dpatchigolla) [inline src/predictions/profiles_mlcorelib/py_native/llm.py:220]: You can keep eligible users as empty string, and keep this where condition based on that. 
ex:
```
filtered_entity_var_str = f"where {eligible_users_prompt_replaced}" if eligible_users_prompt_replaced else ""
filtered_entity_var_table_cte = f"with filtered_entity_var_table as (select * from entity_var_table {filtered_entity_var_str})"


```
And use the string `filtered_entity_var_table_cte` in this query

**Review by dpatchigolla — APPROVED** [review]

File: src/predictions/profiles_mlcorelib/py_native/llm.py (hunk 5 above)
Comment (joker2411) [reply src/predictions/profiles_mlcorelib/py_native/llm.py]: added.

**Review by RuLeKiLLeR — APPROVED** [review]

**Pull request merged**